### PR TITLE
Fix run time errors with python3.12 and pyside6

### PR DESCRIPTION
**Pull request by dov**
Trying to run the Draft benchbase from the main branh generated several python errors. Some probing showed that the sources needed fixes for PySide6 and python3.12.

The two fixes in this pull request are:

1. The `io` module in python3.12 has been renamed `_io`. and therefore required the pyopen assignment to be changed to:
```patch
-if open.__module__ in ['__builtin__', 'io']:
+if open.__module__ in ['__builtin__', 'io', '_io']:
```
2. QtFont.setWeight now takes an argument of like `QtGui.QFont.Weight.Bold`

### CODE REVIEW

File: src/Mod/Arch/exportIFC.py
Comment (lyphrowny) [inline src/Mod/Arch/exportIFC.py:58]: I would suggest using tuples instead of lists. Tuples are faster to create, faster to check for membership, and take up less space. (although these are all minor differences for a list of 3 elements, I'd still consider it good practice)
```suggestion
if open.__module__ in ('__builtin__', 'io', '_io'):
```

**Review by lyphrowny — COMMENTED** [review]
Could you please add the following lines to the FEM WB? The files concerned are `import*` in the following directory: https://github.com/FreeCAD/FreeCAD/tree/ee39814f0c1ae96de446ac96070c9d0eb277128f/src/Mod/Fem/feminout

___

#### I'll leave it here for now, although it might be worth a separate issue:
BTW, is there another way to do this besides shadowing the built-in name? 

This kind of code is repeated in multiple WBs and in multiple files within each WB. This leads to refactoring mistakes like the one in this commit: https://github.com/FreeCAD/FreeCAD/commit/3137dea94044abb4e4bcb43165358c940bb425e2#diff-72525809681e70eb14554b4704983f48437db145786cee2a642a09265690549a - the FEM WB has been working with a redefined `open` for more than a year! (of course this is only the case when reloading `import*` file, but still) 

Or maybe change the name, FreeCAD opens the files with: instead of `open`, use `fc_open`?

File: src/Mod/Draft/DraftGui.py
Comment (lyphrowny) [inline src/Mod/Draft/DraftGui.py:134]: In the documentation for both [Qt5](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QWidget.html#PySide2.QtWidgets.PySide2.QtWidgets.QWidget.close) and [Qt6](https://doc.qt.io/qtforpython-6/PySide6/QtWidgets/QWidget.html#PySide6.QtWidgets.QWidget.close) I find that the return value of `.Close` is bool, but bool doesn't have a `.value` attribute. Am I missing something here?

**Review by lyphrowny — CHANGES_REQUESTED** [review]
I've delved deeper into the shadowing of `open` and found what I believe to be a more correct alternative (If you agree with the proposed changes, and it is too much of a burden to apply them, I can make a PR to your branch with these changes)

I've read the Qt5 and Qt6 docs regarding the proposed changes to Qt and have questions about them

I haven't tested the PR

File: src/Mod/Arch/exportIFC.py
Comment (dov) [reply src/Mod/Arch/exportIFC.py]: Ok. Great catch. Did it

File: src/Mod/Arch/importIFClegacy.py
Comment (dov) [reply src/Mod/Arch/importIFClegacy.py:1799]: Ok. Got rid of it.

File: src/Mod/Draft/DraftGui.py
Comment (dov) [reply src/Mod/Draft/DraftGui.py:134]: I can't explain it yet, but this is what I get
```
$ ipython                                                         
Python 3.12.2 (main, Feb 21 2024, 00:00:00) [GCC 13.2.1 20231205 (Red Hat 13.2.1-6)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.22.2 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import PySide6.QtWidgets as QtWidgets

In [2]: int(QtWidgets.QDialogButtonBox.Close)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 int(QtWidgets.QDialogButtonBox.Close)

TypeError: int() argument must be a string, a bytes-like object or a real number, not 'StandardButton'

In [3]: QtWidgets.QDialogButtonBox.Close.value
Out[3]: 2097152

In [6]: import PySide6
In [7]: PySide6.__version__
Out[7]: '6.6.2'
```

**Comment by dov** [comment]
I found another python problem which is the missing raw-escaping of regular expression strings. Should I push a fix here as well, or should I open another issue.

Search for `re.findall(` in the repo. Here is an example of a such a warning:

```
src/Mod/Arch/ArchSite.py
95       buf = re.findall("point \[(.*?)\]",buf)
```
This should be r-quoted
```
 buf = re.findall(r"point \[(.*?)\]",buf)
```

File: src/Mod/Draft/DraftGui.py
Comment (lyphrowny) [reply src/Mod/Draft/DraftGui.py:134]: I was looking at the wrong places in the documentation, here are the correct ones [Qt5 ](https://doc.qt.io/qtforpython-5/PySide2/QtWidgets/QDialogButtonBox.html#PySide2.QtWidgets.PySide2.QtWidgets.QDialogButtonBox.StandardButton) and [Qt6](https://doc.qt.io/qtforpython-6.2/PySide6/QtWidgets/QDialogButtonBox.html#PySide6.QtWidgets.PySide6.QtWidgets.QDialogButtonBox.StandardButton). In Qt5 `StandardButton` is some class, that has `__int__` method, while in Qt6 `StandardButton` is `enum.Flag`, hence the need for `.value`

File: src/Mod/Arch/exportIFC.py
Comment (lyphrowny) [inline src/Mod/Arch/exportIFC.py:59]:
```suggestion
from builtins import open as pyopen
```

File: src/Mod/Draft/importDWG.py
Comment (lyphrowny) [inline src/Mod/Draft/importDWG.py:55]: Not used in the file, remove?

File: src/Mod/Arch/importIFClegacy.py
Comment (lyphrowny) [reply src/Mod/Arch/importIFClegacy.py:1799]: Not yet :) `bold.setWeight` is still here

File: src/Mod/Draft/DraftGui.py
Comment (lyphrowny) [inline src/Mod/Draft/DraftGui.py:315]: Same here

File: src/Mod/Draft/DraftGui.py
Comment (lyphrowny) [reply src/Mod/Draft/DraftGui.py:134]: We might want to be compatible with Qt5 here
```py
try:
    # pyside2
    return int(QtWidgets.QDialogButtonBox.Close)
except TypeError:
    # pyside6
    return QtWidgets.QDialogButtonBox.Close.value
```

**Review by lyphrowny — CHANGES_REQUESTED** [review]
Could you please move the `from builtins import ...` to the top to other imports? It's a module-level import and seeing it in the middle of the files feels strange

Please remove the `bold.setWeight` as it's not yet removed

Thank you for contribution!

**Comment by lyphrowny** [comment]
> Should I push a fix here as well, or should I open another issue?

I would open another issue

**Comment by chennes** [comment]
I've switched this to draft while it's still a work-in-progress -- it looks like there are some real CI failures in there.

File: src/Mod/Arch/exportIFC.py
Comment (dov) [reply src/Mod/Arch/exportIFC.py:59]: Ok, I did both and pushed. Hopefully (we'll see in the CI) there are no more python errors.

**Comment by dov** [comment]
@chennes @lyphrowny I believe I have fixed all the issues mentioned in this discussion, but the CI still fails, but to me I see no relation to my changes. Could you please give me some feedback on what failures you consider "real" and need to be solved before this PR can be merged?

File: src/Mod/CAM/Path/Post/scripts/estlcam_post.py
Comment (lyphrowny) [inline src/Mod/CAM/Path/Post/scripts/estlcam_post.py:161]: This comment is now obsolete

File: src/Mod/Draft/DraftGui.py
Comment (lyphrowny) [reply src/Mod/Draft/DraftGui.py:134]: We might want to be Qt5 compatible here

```suggestion
        try:
            # pyside2
            return int(QtWidgets.QDialogButtonBox.Close)
        except TypeError:
            # pyside6
            return QtWidgets.QDialogButtonBox.value
```

File: src/Mod/Draft/importDXF.py
Comment (lyphrowny) [inline src/Mod/Draft/importDXF.py:89]: The comment is obsolete

File: src/Mod/Idf/Idf.py
Comment (lyphrowny) [inline src/Mod/Idf/Idf.py:30]: The comment is obsolete

**Review by lyphrowny — CHANGES_REQUESTED** [review]
The tests are failing because of the Materials WB, not the changes introduced in this PR. Update the branch to fix this

Although the tests should pass after the branch update, we should still add Qt5 compatibility to `QButtonBox.Close`, as FreeCAD is still mainly using Qt5 for now

I've added some nitpicks regarding code comments that are now obsolete, could you please fix them as well?

**Review by lyphrowny — APPROVED** [review]
Excellent work! Thank you!

Could you move it out of "Draft"? Otherwise it won't get reviewed when merging on Monday

**Comment by dov** [comment]
> Excellent work! Thank you!
> 
> Could you move it out of "Draft"? Otherwise it won't get reviewed when merging on Monday

Thanks! It was a pleasure, FreeCAD is an amazing program. I'll continue to learn FreeCAD and test the latest version, and will report on any issues I encounter. I was dismayed that FreeCAD was dropped from Fedora 39, and I hope that this work will help in returning it back.

File: src/Mod/Arch/exportIFC.py
Comment (adrianinsaval) [inline src/Mod/Arch/exportIFC.py:60]: the new import replacing this is missing

File: src/Mod/Draft/DraftGui.py
Comment (adrianinsaval) [inline src/Mod/Draft/DraftGui.py:140]: this seems like a separate problem and I would like to have this addressed separately to avoid conflicts with other PRs, see https://github.com/FreeCAD/FreeCAD/pull/13611/files#diff-fdf0c359ca3d9da6b4c4614e349806528f209f38894ad323977bc78da2f1ccbe

File: src/Mod/CAM/Path/Post/scripts/grbl_post.py
Comment (adrianinsaval) [inline src/Mod/CAM/Path/Post/scripts/grbl_post.py:411]: I don't particularly mind it but why change this here but not in the rest?

File: src/Mod/Arch/exportIFC.py
Comment (dov) [reply src/Mod/Arch/exportIFC.py:60]: Thanks. I pushed a fix. The import was missing from `importGBXML.py`

File: src/Mod/CAM/Path/Post/scripts/grbl_post.py
Comment (dov) [reply src/Mod/CAM/Path/Post/scripts/grbl_post.py:411]: I have to admit that I'm not familiar enough with the source to realize that the same pattern appears in other files as well. Please let me know if you want me to revert it and open another pull request which fixes it globally.

File: src/Mod/Draft/DraftGui.py
Comment (adrianinsaval) [reply src/Mod/Draft/DraftGui.py]: Thanks, but I meant to leave this as it is in main before this PR as it is also being touched by another PR that is pending review

File: src/Mod/Draft/DraftGui.py
Comment (adrianinsaval) [inline src/Mod/Draft/DraftGui.py:134]:
```suggestion
        return int(QtWidgets.QDialogButtonBox.Close)
```

File: src/Mod/Draft/DraftGui.py
Comment (adrianinsaval) [reply src/Mod/Draft/DraftGui.py]: took care of it, will merge once CI passes

**Comment by dov** [comment]
🥳

**Comment by ItsAymn** [comment]
Hi there,
This might be a question with a straightforward answer but where do i change the pyopen assignment?
Have been searching for it for a while but have never been able to find a path for it. Asking since I am encountering the issue using CfdOF on Ubuntu 24.04

**Comment by davids91** [comment]
Does anyone know when will this be included in the official Arch Repo package?

**Pull request merged**
